src/battle_script_commands.c

```diff
@@ -5349,14 +5349,14 @@ static bool32 TryKnockOffBattleScript(u32 battlerDef)
     return FALSE;
 }
 
-#define SYMBIOSIS_CHECK(battler, ally)                                                                                               \
-    GetBattlerAbility(ally) == ABILITY_SYMBIOSIS                   \
-    && gBattleMons[battler].item == ITEM_NONE                      \
+#define SYMBIOSIS_CHECK(battler, ally)                             \
+    (gBattleMons[battler].item == ITEM_NONE                        \
     && gBattleMons[ally].item != ITEM_NONE                         \
+    && gBattleMons[battler].hp != 0                                \
+    && gBattleMons[ally].hp != 0                                   \
     && CanBattlerGetOrLoseItem(battler, gBattleMons[ally].item)    \
     && CanBattlerGetOrLoseItem(ally, gBattleMons[ally].item)       \
-    && gBattleMons[battler].hp != 0                                \
-    && gBattleMons[ally].hp != 0
+    && GetBattlerAbility(ally) == ABILITY_SYMBIOSIS)
```

Comment:
Also ghoul mentioned once the it's faster to use `AI_DATA->abilities[ally]`

Reply:
> why is it better to check ability last instead of first?

Because it needs all conditions to pass, if one of the previous conditions do not pass, it doesn't even check it.

> Also ghoul mentioned once the it's faster to use `AI_DATA->abilities[ally]`

Now that's a different optimization that I was unfamiliar with xD
(I can add it if you want)

Reply:
> Because it needs all conditions to pass. If one of the previous conditions do not pass, it doesn't even check it.
I understand but aren't all other conditions more common so they will all be checked until the ability isn't Symbiosis?

> I can add it if you want
yes please

Reply:
Sorry my bad. It's in `battle_script_commands`. `AI_DATA` is only used for ai calcs.

Reply:
> I understand but aren't all other conditions more common so they will all be checked until the ability isn't Symbiosis?

It doesn't matter that they're more common. They all need to pass to enter the if block, so it's better that it first checks the stuff that uses less cycles to check, like direcly reading the value of a stored variable/struct vs calling a multilayered function such as `GetBattlerAbility`.

To put it into perspective, `GetBattlerAbility` does the following:
- Checks if it's being supressed by Gastro Acid.
- Checks if Neutralizing Gas is in the field which means:
  - Checking that the battler is alive.
  - Checking that they actually have the ability.
  - Checking that they're not affected by Gastro Acid.them has their ability supressed by Gastro Acid
- Checks if Mycelium Might is in the field which means:
  - Checking that the battler is alive.
  - Checking that they actually have the ability.
  - Checking that the current move is a status move.
- Checks that the ability is a type of Mold Breaker and:
  - It's not affected by Gastro Acid or it's a move that ignores ability
  - That the ability can be ignored by Mold Breaker
  - That it only ignores the ability of the target and not the attacker
  - That the attacker is using a move.

We can skip all these checks if we first check the items of the battlers, then their HP and if they can lose/get items.

Reply:
I see

Reply:
> > understand but aren't all other conditions more common so they will all be checked until the ability isn't Symbiosis?
>
> It doesn't matter that they're more common.

I think Alex might be right that we should check the most common things first. The logic being, let's say it takes 50 cycles to check uncommonly-rejected things, and 200 cycles to check commonly-rejected things.

1. If you get into the `if` it always took 250 cycles.
2. If it's `common && uncommon` then it usually takes 200 cycles to skip the `if`.
3. If it's `uncommon && common` then it usually takes 250 cycles to skip the `if`.

Of course in case 3, if you have a rare case that gets rejected by `uncommon` then it takes 50 cycles to skip the `if`.

idt it really matters that much, but I think Alex's logic is sound and perhaps we could even apply this logic to `GetBattlerAbility` itself, if there's some way to reorder to make it more likely to return faster?

Reply:
fwiw I haven't given any thought to the specific situation described here, I'm just defending Alex's question :)

Reply:
I think in this case Edu is probably right. `GetBattlerAbility` is much more involved then all the other checks so it would take the most cycles. Though that might also be a reason for putting it first? Since all of the other checks are more common `GetBattlerAbility` will be run almost every-time regardless of it's place. So it might make sense to check it first?

Reply:
So, I really, really don't think it matters too much. But in concept you want to order an `&&` chain so that the most likely to be `FALSE` comes first (to break out of the chain ASAP), and an `||` chain so that the most likely to be `TRUE` comes first (again, to break out of the chain ASAP).

I certainly wouldn't want to sacrifice readability for performance unless we'd measured and found that it made a big enough difference that the readability hit was worth it. In cases where it's equally readable either way, I would lean towards the rule of thumb I gave above.

Reply:
> So, I really, really don't think it matters too much. But in concept you want to order an && chain so that the most likely to be FALSE comes first (to break out of the chain ASAP), and an || chain so that the most likely to be TRUE comes first (again, to break out of the chain ASAP).

Yep, this was my reasoning for asking the question since the ability check will be false in 99% of cases.